doc/api/external_formats.md

```diff
@@ -0,0 +1,78 @@
+
+<!--introduced_in=REPLACEME-->
+
+> Stability: 1 - Experimental
+
+## Introduction
+
+Node natively understands a handful of formats (see the table in [Modules: Load hook][Load
+customization hook]). Non-native / external formats require a translator (under the hood, this is a
+custom loader). When attempting to run a module with an external format (ex `node main.ts`), node
+will try to detect what it is. At current, when node is able to identify the format, it prints a
+message to stdout with instructions (that lead here).
+
+## Setting up a module translator
+
+There are a few, currently manual, steps you need to complete. We intend to automate these in
+future.
+
+1. You will need to find and install a module translator via your preferred package manager. Some
+   package managers provide a CLI utility as well as a website.
+1. Once installed, in order to get node to automatically use it, you will need to create a `.env`
+   file containing a [`--import`][CLI import] for your chosen module translator's package (ex `NODE_OPTIONS="--import=example-translator/register"`). See [CLI:
+   env file][CLI env file].
+1. When you subsequently run node, include the env file flag (ex `node --env-file=.env main.ts`).
+
+## Publishing a module translator
+
+A few specific things are necessary when publishing a module translator.
+
+### Package registration
+
+Your package registration needs the following tags (aka keywords): `module translator`, `node`, and
+the offical name(s) of whatever external format(s) (ex `typescript`) your package handles. We plan
+for node to eventually search by these tags automatically when a user tries to run an unhandled
+external formal; if your package is not registered with the tags, it will not be included in the
+list of suggestions presented to the user.
+
+### Package setup
+
+Your package exports must contain a `"register"` key and should also contain keys for the external format(s) it handles, like:
+
+```json
+{
+  "name": "example-translator",
+  "exports": {
+    "css": "./css.mjs",
+    "register": "registration.mjs",
+    "typescript": "./typescript.mjs"
+  }
+}
+```
+
+The filenames (the values in `"exports"`) can be whatever you want (see [package.json entry
+points][Package entry points]).
+
+`css.mjs` and `typescript.mjs` from the above example are [custom loaders][Module customization hooks]. They will need a [`resolve`][Resolve customization hook] hook that sets `format` to the
+format it handles (ex `'css'` and `'typescript'` respectively), a [`load`][Load customization hook] hook
+that translates the source of the external format to something node understands, and optionally an
+[`initialize`][Initialize customization hook] hook.
+
+`registration.mjs` registers your loader(s) with node (see [module.register][Module register]). It would look something like:
+
+```mjs
+import { register } from 'node:module';
+
+register('example-translator/css');
+register('example-translator/typescript');
```

Comment:
```suggestion
register('./hooks/css.mjs', import.meta.url);
register('./hooks/typescript.mjs', import.meta.url);
```

Reply:
The whole package will likely be solely the hook(s), so I think this is not a realistic example 🤔

Also, why do it with a relative path + `import.meta.url` instead of exports? Exports achieves the same result but far simpler.

Reply:
Why would the whole package likely be solely the hooks? We’ve already heard from the `ts-node` team and others that they need code on the main thread as well as the hooks thread. If that’s required for TypeScript, it’s likely to be the case for many other external formats as well. We should establish a best practice that’s flexible enough to support all libraries that need to fit into our template.

Reply:
Ah, good point.

Reply:
Because it's not very verbose, and it's better than telling people to add new exports. Many people don't use `exports`. And some authors will want to add this to a larger package like `ts-node` rather than create a package just for this, and they might have their own patterns already for exports.

Reply:
Hmm. I think there are pros and cons to each. Unless you feel strongly, I prefer mine for the examples in the docs because it's easier to grok, and grokability is critical for docs examples. Also, the `import.meta.url` bit we already agree is a gotcha (but a currently necessary evil).

An author of ts-node will know the difference.

Reply:
The docs currently always use `register` with `import.meta.url` I believe, because it's often a mistake to do otherwise.

Reply:
Here it's not though, and it would merely add noise to the example. Exposing `typescript` via `exports` could have additional utility too.

Reply:
If you look at https://nodejs.org/api/module.html#customization-hooks, all of the examples using `register` include `import.meta.url`. Someone can always provide exports _and_ use `import.meta.url`. We should include `import.meta.url` in the example, because it only helps and it’s the best practice we’re trying to encourage for when people use `register`. I think most authors are much more familiar with referencing a file relative to the current file than they are referencing a file based on an `exports` map entry.